### PR TITLE
Issue 350: Optimise stan via profiling

### PR DESCRIPTION
**Pull request by seabbs**
## Description

This PR  is related to #350 but doesn't close it as there is more work to do. 

Using profiling I have identified areas that are bottlenecking the code (mostly `expected_obs`) and performance tuned with the slight cost of readability. 

I have also looked at the trajectory and posteriors and identified that our diffuse prior on the overdispersion may lead to some significant slow down. Given we have low levels of belief in most settings I propose we moderately tigthen it. In terms of posterior I have seen no impact of this. 

I have also gone through all the case studies and examples and checked the `adapt_delta` and `max_treedepth` settings. In most cases I have been able to relax these with no diagnostic issues.

Tagged a few reviewers who may be interested in taking a look. @zsusswein may also be interested. @jessalynnsebastian I pinged you as I edited some of the tests and docs for `expected_obs()` in this PR which you recently touched.

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title in the for Issue(s) *issue-numbers*: PR title
- [x] I have read the [contribution guidelines](https://github.com/epinowcast/.github/blob/main/CONTRIBUTING.md).
- [x] I have tested my changes locally.
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required.
- [x] My code follows the established coding standards.
- [x] I have added a news item linked to this PR.
- [x] I have reviewed CI checks for this PR and addressed them as far as I am able.

### CODE REVIEW

File: inst/examples/germany_parametric_with_non_parametric_effect.R [hunk 1]
```diff
@@ -93,7 +93,7 @@ expectation_module <- enw_expectation(
 # may not be the case
 reference_module <- enw_reference(
   parametric = ~ 1,
-  non_parametric = ~ 1 + delay0mod,
+  non_parametric = ~ 0 + delay0mod,
```
Comment (seabbs) [inline inst/examples/germany_parametric_with_non_parametric_effect.R:96]: This was previously mistakenly updated and I included the fix here. I don't really think its worth putting in its own PR...

File: inst/stan/functions/expected_obs.stan [hunk 2]
```diff
@@ -38,79 +40,79 @@
...
+ *               rep(0.000, 21))
  *
  * Example with aggregation of probabilities where aggregation occurs on every fifth day
  * agg_probs <- matrix(c(rep(0, times = 30 * 4),
  *                       rep(c(rep(1, times = 5),
  *                             rep(0, times = 6 * 5 + 30 * 4)),
  *                           times = 5),
  *                       rep(1, times = 5)), ncol = 30, byrow = TRUE)
- * eobs <- exp(expected_obs(tar_obs, date_p + rep(0, 30), 1, 1, agg_probs))
+ * eobs <- exp(expected_obs(tar_obs, date_p + rep(0, 30), 30, 1, agg_probs))
```
Comment (jessalynnsebastian) [inline inst/stan/functions/expected_obs.stan:94]: Think this example is now missing one of the indicators - ref_as_p/agg_probs

```suggestion
 * eobs <- exp(expected_obs(tar_obs, date_p + rep(0, 30), 30, 1, 1, agg_probs))
```

File: tests/testthat/test-stan_expected_obs.R [hunk 3]
```diff
@@ -3,11 +3,66 @@ skip_on_os("windows")
 skip_on_os("mac")
 skip_on_local()
 
-enw_stan_to_r(
-  c("expected_obs.stan", "hazard.stan"),
-  system.file("stan", "functions", package = "epinowcast")
-)
-# note these tests require enw_stan_to_r() to be run first
```
Comment (jessalynnsebastian) [inline tests/testthat/test-stan_expected_obs.R:10]: Why remove this bit?

File: inst/stan/functions/expected_obs.stan (hunk 2 above)
Comment (seabbs) [reply inst/stan/functions/expected_obs.stan:94]: Good catch!

File: tests/testthat/test-stan_expected_obs.R (hunk 3 above)
Comment (jessalynnsebastian) [reply tests/testthat/test-stan_expected_obs.R:10]: In this case should we also remove it from the discretized logit hazard test file?

Comment (seabbs) [reply tests/testthat/test-stan_expected_obs.R:10]: yes good spot

**Review by jessalynnsebastian — APPROVED** [review]
All looks good to me after those minor changes

**Comment by seabbs** [comment]
Going to monitor this to see if we see any performance impacts from the change in phi prior

**Pull request merged**